cmd/shadowsocks-server/server.go

```diff
@@ -451,5 +486,113 @@ func main() {
 		}
 	}
 
+	if managerAddr != "" {
+		addr, err := net.ResolveUDPAddr("udp", managerAddr)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Can't resolve address: ", err)
+			os.Exit(1)
+		}
+		conn, err := net.ListenUDP("udp", addr)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Error listening:", err)
+			os.Exit(1)
+		}
+		log.Printf("manager listening udp addr %v ...\n", managerAddr)
+		defer conn.Close()
+		go managerDaemon(conn)
+	}
+
 	waitSignal()
 }
+
+func managerDaemon(conn *net.UDPConn) {
+	for {
+		data := make([]byte, 300)
+		_, remote, err := conn.ReadFromUDP(data)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Failed to read UDP manage msg, error: ", err.Error())
+			continue
+		}
+		command := string(data)
+		var res []byte
+		switch {
+		case strings.HasPrefix(command, "add:"):
+			res = handleAddPort(bytes.Trim(data[4:], "\x00\r\n "))
+		case strings.HasPrefix(command, "remove:"):
+			res = handleRemovePort(bytes.Trim(data[7:], "\x00\r\n "))
+		case strings.HasPrefix(command, "ping"):
+			res = handlePing()
+		}
+		if len(res) == 0 {
+			continue
+		}
+		_, err = conn.WriteToUDP(res, remote)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Failed to write UDP manage msg, error: ", err.Error())
+			continue
+		}
+	}
+}
+
+func handleAddPort(payload []byte) []byte {
+	var params struct {
+		ServerPort interface{} `json:"server_port"` // may be string or int
+		Password   string      `json:"password"`
+	}
+	json.Unmarshal(payload, &params)
+	if params.ServerPort == nil || params.Password == "" {
+		fmt.Fprintln(os.Stderr, "Failed to parse add req: ", string(payload))
+		return []byte("err")
+	}
+	port := parsePortNum(params.ServerPort)
+	if port == "" {
+		return []byte("err")
+	}
+	passwdManager.updatePortPasswd(port, params.Password, config.Auth)
+	return []byte("ok")
+}
+
+func handleRemovePort(payload []byte) []byte {
+	var params struct {
+		ServerPort interface{} `json:"server_port"` // may be string or int
+	}
+	json.Unmarshal(payload, &params)
+	if params.ServerPort == nil {
+		fmt.Fprintln(os.Stderr, "Failed to parse remove req: ", string(payload))
+		return []byte("err")
+	}
+	port := parsePortNum(params.ServerPort)
+	if port == "" {
+		return []byte("err")
+	}
+	log.Printf("closing port %s\n", port)
+	passwdManager.del(port)
+	return []byte("ok")
+}
+
+func handlePing() []byte {
```

Comment:
handlePing function must return 'pong' as the management protocol

Reply:
aha, I see that